### DISCUSSION TITLE
differences between connectOrCreate and upsert?

### DISCUSSION THREAD

**Discussion by johhansantana** [prisma/prisma#3024] [Q&A] [answered]
In https://github.com/prisma/prisma/releases/tag/2.1.0 release, you can use `connectOrCreate`.

aren't they the same thing?

**Comment by nikolasburk** [comment, marked as answer]
Hey @johhansantana, thanks a lot for raising this!

So, `connectOrCreate` is used as a _nested_ operation whereas `upsert` is a top-level operation. We also chose the wording for `connectOrCreate` in a way that it's intuitively clear what's happening: Fist we try to _connect_ to the record specified by `where`. If the record does not exist, we _create_ a new one. Functionality-wise, they're indeed similar but we chose a bit more precise wording for this one to make it easier to understand.

On a technical level, the two are also different and take different input arguments:

```ts
await prisma.user.update({
  data: {
    name: 'Alice',
    posts: {
      upsert: {
        create: { title: 'Hello World' },
        where: { id: 1 },
        update: { title: 'Hello World' },
      },
    },
  },
  where: { id: 42 },
})
```

```ts
await prisma.user.update({
  data: {
    name: 'Alice',
    posts: {
      connectOrCreate: {
        create: { title: 'Hello World' },
        where: { id: 1 },
      },
    },
  },
  where: { id: 42 },
})
```

Does that help? 🙂

**Reply by johhansantana** [reply]
oh, so instead of updating the posts it will try to connect to one of the posts if it already exists or create a new one if it doesn't, it won't update the posts at all.

I think I get it now. Thanks!

**Reply by nikolasburk** [reply]
Yeah, admittedly the difference is not super obvious until you have a real use case for it 🙂

**Reply by ctsstc** [reply]
I find this example super helpful. I am a bit confused though about the direction of the relationships maybe since I'm having issues getting this to work. 

For example if I have employment history with an employer, and I would like to connect or create the employer for the employment history I get auto complete for this code, but then when it goes to run it I'll get this error:

## Error
```
Unknown arg `employer` in data.employer for type EmploymentHistoryUncheckedUpdateInput. Did you mean `employerId`?
```

## Example Schema
```prisma
model Employer {
  id                        String                  @id @default(uuid()) @db.Uuid
  name                      String                  @db.VarChar
  employmentHistory         EmploymentHistory[]
}

model EmploymentHistory {
  id                String          @id @default(uuid()) @db.Uuid
  employerId        String          @db.Uuid
  employer          Employer        @relation(fields: [employerId], references: [id])
}
```

## Example Code
```ts
this.prisma.employmentHistory.update({
  where: {
    id: employmentHistoryId
  },
  data: {
    employer: {
      connectOrCreate: {
        where: {
          id: employerId
        },
        create: {
          name: employerName
        }
      }
    }
  }
})
```

I was really hoping this `connectOrCreate` would work. I also tried looking into the `upsert`, but it didn't allow me to put a `where` clause on it, I assume because of the relationship -- does it implicitly look for the connection instead of allowing the where, or is this related to nesting?

## Conclusion

I would like to better understand the above doesn't work.

Am I stuck looking up the `employerId` via a separate upsert query?

## Example Double Query Lookup for Relationship

### Example 1 - Condensed

```ts
this.prisma.employmentHistory.update({
  where: {
    id: employmentHistoryId,
  },
  data: {
    employerId: (
      await this.prisma.employer.upsert({
        where: {
          id: employerId,
        },
        create: {
          name: employerName,
        },
        update: {},
      })
    ).id,
  }
});
```

### Example 2 - Separated

```ts
const employerId = (
  await this.prisma.employer.upsert({
    where: {
      id: employerId,
    },
    create: {
      name: employerName,
    },
    update: {},
  })
).id

this.prisma.employmentHistory.update({
  where: {
    id: employmentHistoryId,
  },
  data: {
    employerId,
  }
});
```

It would be nice to have this as a single query but maybe it's not possible?

I'm assuming if I was updating an employer I could connect or create employment history, but going the other direction doesn't work? I did check the types and it does not have the `employer` key on `EmploymentHistoryUncheckedUpdateInput`, but I'm not sure why it was letting me autocomplete this initially as if it were fine (maybe some crazy typing permutation or better to display it than not at all).

**Reply by kaufmann42** [reply]
+1 The above question is asked all over the internet. Why doesn't connectOrCreate work properly and what is this mysterious error message?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
